### ISSUE TITLE
Code review - Susanfe

### ISSUE THREAD

**Issue by mochiruntime** [Susanfe/sdp-event-management#249]
This is a code review for the code of the whole project on master branch at commit 
https://github.com/Susanfe/sdp-event-management/tree/86b29d350c4885af957dea7a37a55f54ab37ed60

Of course, we don't expect you to change all of your code (some parts of the review can be subjective or arguable anyway), important is that you understand why better alternatives exist.

**Please ask us if you have any question about the code review, whether it is by writing a comment below or by sending us an email or meeting us.**

**Important : if you try to fix these, don't break your app. I prefer that the app works even if you don't manage to fix anything. It may be too hard to rework many of the issues below, as it would require to rewrite the whole architecture.**

Globally, your architecture is very clean and uses a modern toolset which is nice. It almost feels a bit over-engineered compared to the size of your app, but this is great for scaling and maintenance. Your code is mostly good, however there are a few important concepts missing, please read the whole review.

## Packages

You make good use of packages with regular naming conventions.

## Notifications

Cool use of the strategy pattern for the different notification schedules with a typed implementation for each kind.
Your SchedulerHelper nicely isolates the Android implementation details and abstracts them away at the same time. Since you mainly use it as a utility class providing static methods, you could mark the class final and make the constructor private to make sure that noone can misuse the class.

## Models

Your classes are pretty good in general. I have a few concerns, which mainly boil down to implementation choices : for a few of those points, I may not know how your classes interact with the whole codebase so you may have valid reasons for your implementation :

- `DummyUser` and `DummyInMemorySession` : can't these classes be in your test packages since you only use them for testing purposes ? You have development logic into your production code, which isn't ideal (it bloats the release size and can confuse other developers as it breaks the separation of concerns provided by the packages).

- Some classes don't have Javadoc : try to adopt good habits and write the doc as you write code, it will help your colleagues and your future self.

- Your classes are mutable (which is not a problem by itself, this is more of an open discussion). However, using immutability would allow you to get rid of state which often eliminates many nasty bugs and maintenance issues. Of course, this is a trade-off against higher memory usage and more garbage collection, but many modern front-end frameworks are built around it because it allows you to express the view for any state, avoiding the tedious work of manually changing the view components. In your implementation, you have many empty constructors for which you then use setters to fill the fields, but for many of them you could collect the data beforehand and use a complete constructor.

- ### [Important] No abstraction for external libraries 

    Your current code is missing many abstractions between your classes and the external libraries you use, such as Firebase or Glide. I understand that your Repository classes kind of separate the concerns of the Database with the models, but it doesn't provide abstraction for it and it still splits the code. Also, many of your class methods directly call the Firebase / Glide API.
    
    This is a huge problem for maintainability : what would happen if your product owner suddenly decides that you need to use a relational database, graph-oriented database or another image loading library? You would then need to modify many classes (such as Event, FirebaseBackedUser, Repository classes), everywhere you used the proprietary Firebase API or Glide API directly.

    The solution to this is regrouping common library logic into a single class : you could have an interface Database (respectively ImageLoader) that only describes the actions you want to perform on the database (images) without any implementation logic. For instance :

    ```java
        public interface Database {
            // i'm using futures here but you could also have callbacks for Java < 8
            public Future<List<Event>> getAllEvents();
            public Future<User> getUserById(String uid);
            ... // you get the idea
        }

        public interface ImageLoader {
            public <S> void display(S source, ImageView view, List<BitmapTransformation> transforms);
        }
    ```

    Then, you could have concrete definitions for these interfaces that actually implement the proprietary logic (it may make sense to use singletons).

    ```java
        public class FirebaseDatabase implements Database {
            FirebaseDatabase db = FirebaseDatabase.getInstance();
            StorageReference store = FirebaseStorage.getInstance();

            public Future<List<Event>> getAllEvents() {
                Future<List<Event>> f = new CompletableFuture<List<Event>>();
                // i don't know your exact firebase structure so this is approximate
                db.getReference("events").addValueEventListener(new ValueEventListeenr() {
                    @Override
                    public void onDataChange(DataSnapshot data) {
                        List<Event> events = new ArrayList<>();
                        for (DataSnapshot event : snapshot.getChildren()) {
                            events.add(event.getValue(Event.class));
                        }
                        f.complete(events);
                    }
                    ...
                });
                return f;
            }

            ... // all your other database accesses defined
        }
    ```

    Finally, you instantiate these once and pass them around using dependeny injection. Each of your fragment / activity will receive a `Database` object (which underlying instance would be a `FirebaseDatabase`), and you can perform your actions without writing any implementation logic at call site.

    This is very nice, because let's say you want to suddenly change to a relational database now : then you only need to create a single new `SQLDatabase` class which implements `Database`. Then, instead of instantiating a FirebaseDatabase, you create a SQLDatabase. Voilà, no other changes and it works out of the box. The same procedure applies to ImageLoader.

#### Event.java

- There's probably an error in the Javadoc : shouldn't field users be `Map<String, List<String>>`, or is it a User to Role map? I guess the latter since you have getPermissions which reverses the map. Also could you have typed roles instead of strings?
- This class has an empty constructor, shouldn't it make more sense to collect all the inputs in EventCreateActivity first and then use a complete constructor, avoiding dangling unset references ? 
- get/setUsers : would it make sense to deep copy the Map ? 
- begin/endDateAsString : why not extract these into a single util method in another util class ? You basically perform the same code and it's not related to the concept of Event itself
- @Exclude on methods : shouldn't @Exclude be used on fields so that Firebase ignore them (I may be wrong here) ? 
- uploadImage : contains Firebase logic to online storage and DB inside, as explained above.
- loadEventImageIntoImageView : contains Glide logic inside, as explained above.

#### Feed.java

- Shouldn't the class be named FacebookPost ? Personally I understand Feed as a stream of many things (such as news).

#### FirebaseBackedUser.java

- This class is completely dependent on Firebase logic. As explained above, shouldn't it make more sense that the Database abstraction returns a built User instead? I get that you use it to update it in real time, but it would be nicer to have an abstraction around this mutability (say a generic type instead of the previosu Future to retrieve a data from Firebase on which you can redefine onUpdate or something).

#### MultiDrawable.java 

- Shouldn't this class be moved into some UI package ? It's not really a data class, rather an utility for views.

#### News.java 

- The empty constructor is completely unused (and I think unuseful too), so this class could already be immutable.

#### NewsOrTweetOrFacebook.java

- I don't really understand this class : is the OR inclusive ? If not, then this should be split into 3 classes (FacebookPost, TweeterPost, News) that inherit from a common parent, say Post (given that you have int TYPE_* constants to describe the type, I guess they are exclusive), then you can use a `List<Post>` instead of `List<NewsOrTweetOrFacebook>` (you are basically int-ly typing your class, which is bad) also the mergeLists method becomes obsolete. Else if the OR IS inclusive, the class may make sense but the type should at least be an enum instead.

#### ScheduledItem.java

- The empty constructor is completely unused (and I think unuseful too), so this class could already be immutable.
- dateAsString, same as for Event.java, could be an util method in another util class (you're repeating this code).

#### SpotType.java

- There's a weird unused import : `import java.security.cert.CertPathValidatorException;`. Use Android Studio's automatic tools to format your code and auto-manage imports.

#### Zone.java

- Unused empty constructor (and unuseful)
- There's no Javadoc or comments.

#### `ticketing` package

- Many of these classes do not have documentation. If this is a library, consider adding a `vendor` package and moving it there.

## Bottom line

Remember that sometimes, simpler is better.

**Comment by zyuiop** [comment]
Hello, thanks for the review. I have a few questions about it, mostly because some of the changes you suggest have already been made (maybe it's because some of the review was started earlier, if so, you can ignore the comments):

- the `DummyInMemorySession` class has already been moved to a test package, the `DummyUser` was forgotten in the main package but must indeed be moved (@zyuiop or @Fnux)
- for the firebase abstraction, I think that what we did one or two weeks ago kind of fixes the problem: we have an interface for each repository, and a firebase specific implementation for each of those interfaces. This seems to work well as we're able to switch the implementation easily in our test classes. Tell me if I'm wrong and if this separation is not sufficient.
- for the data-classes, I'm pretty sure the no-args constructor is used by Firebase internally, which disables us from removing it. If not, then we can indeed remove it.
- for @exclude on methods: firebase seems to use a weird mix between method names and field names to determine what data to store, and so we need to annotate the methods themselves 

For the rest, here are some of the updates we should quickly do:
- for the rest of the FB/Glide logic that was added inside event: @zyuiop @vquelque (shouldn't have been merged this way, sorry if I didn't review in time)
- NewsOrTweet: indeed, I don't know why I didn't think about it. @zyuiop 
- FirebaseBackedUser: I agree @Fnux 
- Ticketing: I'll handle it @zyuiop

**Comment by Luminilion** [comment]
Thanks for the review !

- `Feed.java` refactoring @stanjouven 
- `SpotType.java` & `Zone.java` @zbirobin @stanjouven 

In my opinion `NewsOrTweetOrFacebook` could clearly be refactored @Luminilion @stanjouven (@zyuiop but you have a lot of things to do).

**Comment by zyuiop** [comment]
For the former one, I already planned to do it and assigned it to me (#251)

**Comment by mochiruntime** [comment]
> - the `DummyInMemorySession` class has already been moved to a test package, the `DummyUser` was forgotten in the main package but must indeed be moved (@zyuiop or @Fnux)

nice :relaxed: 

> - for the firebase abstraction, I think that what we did one or two weeks ago kind of fixes the problem: we have an interface for each repository, and a firebase specific implementation for each of those interfaces. This seems to work well as we're able to switch the implementation easily in our test classes. Tell me if I'm wrong and if this separation is not sufficient.

I re-reread the implementation, I think it's fine. Note that it is somewhat tightly coupled to the other frameworks you use and the reasoning model they are built upon, so this is a classic example of the compromise you need to make between trusting a library and having your app's logic to be as independent as possible (so there's no single correct solution here).

> - for the data-classes, I'm pretty sure the no-args constructor is used by Firebase internally, which disables us from removing it. If not, then we can indeed remove it.

Ah right I remember Firebase having very specific rules with the way classes have to be written. If you can't remove it, then it's fine.

> - for @exclude on methods: firebase seems to use a weird mix between method names and field names to determine what data to store, and so we need to annotate the methods themselves 

I see, if it's not redundant then it's fine too.
